miriway.cpp

```diff
@@ -66,27 +67,47 @@ std::map<std::string, ShellCommands::CmdFunctor> const wm_command =
         { "exit", [](ShellCommands* sc, bool shift) { sc->exit(shift); } },
     };
 
+struct ShellComponentRunInfo
+{
+    time_t last_run_time = 0;
+    int num_runs_with_too_short_time_in_between = 0;
+    std::unique_ptr<miral::FdHandle> handle = nullptr;
+};
+
+auto constexpr min_time_seconds_allowed_between_runs = 3;
+auto constexpr time_seconds_wait_foreach_run = 3;
+const int max_too_short_reruns_in_a_row = 10;
+
+std::string get_cmd_string(std::vector<std::string> const& cmd)
+{
+    std::ostringstream os;
+    std::copy(cmd.begin(), cmd.end(), std::ostream_iterator<std::string>( os ));
+    return os.str();
+}
+
 // Build a list of shell components from "<commands>" values and launch them all after startup
 struct ShellComponents
 {
-    explicit ShellComponents(std::function<void(std::vector<std::string> const& command_line)> launch) :
+    explicit ShellComponents(std::function<void(std::vector<std::string> const& command_line, ShellComponentRunInfo&)> launch) :
```

Comment:
AFAICS `launch` only takes a `ShellComponentRunInfo` parameter to facilitate populating `run_info`. And I don't think `run_info` should exist. (See other comment about undefined behaviour).